tools/maintenance/maintenance.py

```diff
@@ -88,72 +84,99 @@ class NodeMaintenance:
+        self.per_maint_vms = None
+        self.upc_maint_vms = None
 
     def update_maintenance_nodes(self) -> None:
-        self.maint_nodes = get_maintenance_nodes(self.project, self.regex)
+        self.per_maint_vms = get_maintenance_nodes(self.project)
+        if self.regex:
+            self.per_maint_vms = list(filter(self.regex.search,
+                                           self.per_maint_vms))
+
+        if self.slurm_nodes:
+            self.per_maint_vms = list(set(self.per_maint_vms) &
+                                    set(self.slurm_nodes))
 
     def update_upcoming_maintenance(self) -> None:
-        self.upc_maint = get_upcoming_maintenance(self.project, self.regex)
+        self.upc_maint_vms = get_upcoming_maintenance(self.project)
+        if self.regex:
+            self.upc_maint_vms = list(filter(lambda x: self.regex.match(x[0]),
+                                  self.upc_maint_vms))
+
+        if self.slurm_nodes:
+            self.upc_maint_vms = [u for u in self.upc_maint_vms if u[0] in
+                              self.slurm_nodes]
 
     def print_maintenance_nodes(self) -> None:
-        if self.maint_nodes is None:
+        if self.per_maint_vms is None:
             self.update_maintenance_nodes()
 
-        if not self.maint_nodes:
+        if not self.per_maint_vms:
             print("No nodes with periodic maintenance\n")
             return
 
         print("Nodes with PERIODIC maintenance")
-        for x in self.maint_nodes:
+        for x in self.per_maint_vms:
             print(x)
         print()
 
     def print_upcoming_maintenance(self) -> None:
-        if self.upc_maint is None:
+        if self.upc_maint_vms is None:
             self.update_upcoming_maintenance()
 
-        if not self.upc_maint:
+        if not self.upc_maint_vms:
             print("No upcoming maintenance\n")
             return
 
         print("Upcoming maintenance:")
-        row_format ="{:30}" * (len(self.upc_maint[0]))
+        row_format ="{:30}" * (len(self.upc_maint_vms[0]))
         print(row_format.format(*["Name", "Earliest Start", "Latest Start",
                                   "Can Reschedule", "Maintenance Type"]))
-        for row in self.upc_maint:
+        for row in self.upc_maint_vms:
             print(row_format.format(*row))
         print()
 
 def node_maintenace_factory(project: str, regex: str = None,
-                            run_maintenance: bool = False) -> NodeMaintenance:
-    res = subprocess.run(PRJ_CMD.format(project), shell=True,
-                         capture_output=True, text=True, check=False)
-
-    if res.returncode != 0:
-        err_msg = f"{project} does not exist or you may not have permission" \
-                   " to access it"
-        raise subprocess.SubprocessError(err_msg)
+                            check_maint: bool = False,
+                            slurm: bool = False) -> NodeMaintenance:
+    cmd = PRJ_CMD.format(project)
```

Comment:
Do not implement in this PR, separate PR if you presue.

We could make the project optional and lookup using either `gcloud config get-value project` or `curl http://metadata.google.internal/computeMetadata/v1/project/project-id -H "Metadata-Flavor: Google"`. When called from a VM either of these will produce the project id the vm is on. This would be useful for the slurm case in particular as we are likely calling from a controller or login where nodes will be in the same project. Please open a task for prioritization if you agree.

Reply:
Yeah I had thought about this as well.  It does seem useful, but I'm not sure if it's worth the time right now, especially if we're stopping at giving instructions for a cron task.  The user at that point should have the project available.  If we eventually move to an automatic installation through the use of a module of some kind, at that point I think it would be worth automating the slurm project retrieval.